pkg/apiserver/rest/dualwriter_syncer.go

```diff
@@ -0,0 +1,285 @@
+	ctx, cancelFn := context.WithTimeout(ctx, timeout)
+	defer cancelFn()
+
+	// implementation depends on the current DualWriter mode
+	switch mode {
+	case Mode1, Mode2:
+		return legacyToUnifiedStorageDataSyncer(ctx, mode, legacy, storage, kind, reg, serverLockService, requestInfo)
+	default:
+		klog.Info("data syncer not implemented for mode mode:", mode)
+		return false, nil
+	}
+}
+
+func legacyToUnifiedStorageDataSyncer(ctx context.Context, mode DualWriterMode, legacy LegacyStorage, storage Storage, resource string, reg prometheus.Registerer, serverLockService ServerLockService, requestInfo *request.RequestInfo) (bool, error) {
+	metrics := &dualWriterMetrics{}
+	metrics.init(reg)
+
+	log := klog.NewKlogr().WithName("legacyToUnifiedStorageDataSyncer").WithValues("mode", mode, "resource", resource)
+
+	everythingSynced := false
+	outOfSync := 0
+	syncSuccess := 0
+	syncErr := 0
+
+	maxInterval := dataSyncerInterval + 5*time.Minute
+
+	var errSync error
+	const maxRecordsSync = 1000
+
+	// LockExecuteAndRelease ensures that just a single Grafana server acquires a lock at a time
+	// The parameter 'maxInterval' is a timeout safeguard, if the LastExecution in the
+	// database is older than maxInterval, we will assume the lock as timeouted. The 'maxInterval' parameter should be so long
+	// that is impossible for 2 processes to run at the same time.
+	err := serverLockService.LockExecuteAndRelease(ctx, fmt.Sprintf("legacyToUnifiedStorageDataSyncer-%d-%s", mode, resource), maxInterval, func(context.Context) {
+		log.Info("starting legacyToUnifiedStorageDataSyncer")
+		startSync := time.Now()
+
+		orgId := int64(1)
+
+		ctx = klog.NewContext(ctx, log)
+		ctx = identity.WithRequester(ctx, getSyncRequester(orgId))
+		ctx = request.WithNamespace(ctx, requestInfo.Namespace)
+		ctx = request.WithRequestInfo(ctx, requestInfo)
+
+		storageList, err := getList(ctx, storage, &metainternalversion.ListOptions{
+			Limit: maxRecordsSync,
+		})
+		if err != nil {
+			log.Error(err, "unable to extract list from storage")
+			return
+		}
+
+		if len(storageList) >= maxRecordsSync {
+			errSync = fmt.Errorf("unified storage has more than %d records. Aborting sync", maxRecordsSync)
+			log.Error(errSync, "Unified storage has more records to be synced than allowed")
+			return
+		}
+
+		log.Info("got items from unified storage", "items", len(storageList))
+
+		legacyList, err := getList(ctx, legacy, &metainternalversion.ListOptions{})
+		if err != nil {
+			log.Error(err, "unable to extract list from legacy storage")
+			return
+		}
+		log.Info("got items from legacy storage", "items", len(legacyList))
+
+		itemsByName := map[string]syncItem{}
+		for _, obj := range legacyList {
+			accessor, err := utils.MetaAccessor(obj)
+			if err != nil {
+				log.Error(err, "error retrieving accessor data for object from legacy storage")
+				continue
+			}
+			name := accessor.GetName()
+
+			item, ok := itemsByName[name]
+			if !ok {
+				item = syncItem{}
+			}
```

Comment:
[non-blocking] Retrieving map value for non-existing key returns zero value. Same for the loop on `storageList`.
```suggestion
			item := itemsByName[name]
```